src/or/proto_socks.c

```diff
@@ -116,6 +119,8 @@ parse_socks4_request(const uint8_t *raw_data, socks_request_t *req,
     goto end;
   }
 
+  *drain_out = (size_t)parsed;
```

Comment:
We should assert that this value is >= 0. 

Also, what about the bytes from the socks4a hostname?  I think we should add them to drain_out too.

Reply:
Assert added in bfa4a7da261b4acb75878522d70c305378c1753e. 

Doesn't code from trunnel already include SOCKS4a hostname length into return value? 
```
1648       /* Parse nulterm socks4a_addr_hostname */
1649       {
1650         uint8_t *eos = (uint8_t*)memchr(ptr, 0, remaining);
1651         size_t memlen;
1652         if (eos == NULL)
1653           goto truncated;
1654         trunnel_assert(eos >= ptr);
1655         trunnel_assert((size_t)(eos - ptr) < SIZE_MAX - 1);
1656         memlen = ((size_t)(eos - ptr)) + 1;
1657         if (!(obj->socks4a_addr_hostname = trunnel_malloc(memlen)))
1658           goto fail;
1659         memcpy(obj->socks4a_addr_hostname, ptr, memlen);
1660         remaining -= memlen; ptr += memlen;
1661       }
1662       break;
1663 
1664     default:
1665       break;
1666   }
1667   trunnel_assert(ptr + remaining == input + len_in);
1668   return len_in - remaining;
```